### PR TITLE
CP-1758 Callbacks of the bunch of tags has been added/removed successfully.

### PR DESCRIPTION
**Pull request by Azharhussain20**
- Refactor addSubscriptionTag/removeSubscriptionTag & addSubscriptionTags/removeSubscriptionTags

### CODE REVIEW

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 1]
```diff
@@ -1495,17 +1526,23 @@ + (void)addSubscriptionTag:(NSString*)tagId {
                 if (![subscriptionTags containsObject:tagId]) {
                     [subscriptionTags addObject:tagId];
                 }
+                
                 [userDefaults setObject:subscriptionTags forKey:@"CleverPush_SUBSCRIPTION_TAGS"];
                 [userDefaults synchronize];
+                
+                if(callback){
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1533]: add space:
`if (callback)

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 2]
```diff
@@ -1462,21 +1464,50 @@ + (void)handleJSONNSURLResponse:(NSURLResponse*) response data:(NSData*) data er
...
     [self waitForTrackingConsent:^{
-        NSLog(@"CleverPush: addSubscriptionTag 2: %@", tagId);
         
         NSUserDefaults* userDefaults = [NSUserDefaults standardUserDefaults];
-        __block NSMutableArray* subscriptionTags = [NSMutableArray arrayWithArray:[userDefaults arrayForKey:@"CleverPush_SUBSCRIPTION_TAGS"]];
+        subscriptionTags = [NSMutableArray arrayWithArray:[userDefaults arrayForKey:@"CleverPush_SUBSCRIPTION_TAGS"]];
         
         if ([subscriptionTags containsObject:tagId]) {
-            NSLog(@"CleverPush: addSubscriptionTag - already has tag, skipping API call");
+            if(callback){
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1504]: add space:
`if (callback)

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 3]
```diff
@@ -1516,14 +1553,22 @@ + (void)removeSubscriptionTag:(NSString*)tagId {
...
+                
                 if (!subscriptionTags) {
                     subscriptionTags = [[NSMutableArray alloc] init];
                 }
                 [subscriptionTags removeObject:tagId];
+                
                 [userDefaults setObject:subscriptionTags forKey:@"CleverPush_SUBSCRIPTION_TAGS"];
                 [userDefaults synchronize];
+                
+                if(callback){
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1568]: add space:
`if (callback)

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 4]
```diff
@@ -1462,21 +1464,50 @@ + (void)handleJSONNSURLResponse:(NSURLResponse*) response data:(NSData*) data er
...
++ (void)addSubscriptionTags:(NSArray*)tagIds callBack:(void(^)(NSArray *))callback {
+    dispatch_group_t group = dispatch_group_create();
+    for (NSString* tagId in tagIds) {
+        dispatch_group_enter(group);
+        [self addSubscriptionTag:tagId callback:^(NSString *tagId) {
+            dispatch_group_leave(group);
+        }];
+    }
+    dispatch_group_notify(group, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
+        if(callback){
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1476]: add space:
`if (callback)

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 5]
```diff
@@ -1462,21 +1464,50 @@ + (void)handleJSONNSURLResponse:(NSURLResponse*) response data:(NSData*) data er
...
++ (void)removeSubscriptionTags:(NSArray*)tagIds callBack:(void(^)(NSArray *))callback{
+    dispatch_group_t group = dispatch_group_create();
+    for (NSString* tagId in tagIds) {
+        dispatch_group_enter(group);
+        [self removeSubscriptionTag:tagId callback:^(NSString *tagId) {
+            dispatch_group_leave(group);
+        }];
+    }
+    dispatch_group_notify(group, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
+        if(callback){
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1491]: add space:
`if (callback)

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 6]
```diff
@@ -92,6 +92,7 @@ @implementation CleverPush
 NSMutableArray* pendingChannelConfigListeners;
 NSMutableArray* pendingSubscriptionListeners;
 NSMutableArray* pendingTrackingConsentListeners;
+NSMutableArray* subscriptionTags;
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:95]: Why do we have a global variable here? Please remove it

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 7]
```diff
@@ -249,6 +250,7 @@ + (id)initWithLaunchOptions:(NSDictionary*)launchOptions
     pendingSubscriptionListeners = [[NSMutableArray alloc] init];
     pendingTrackingConsentListeners = [[NSMutableArray alloc] init];
     autoAssignSessionsCounted = [[NSMutableDictionary alloc] init];
+    subscriptionTags = [[NSMutableArray alloc] init];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:253]: Why do we have a global variable here? Please remove it

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 8]
```diff
@@ -309,7 +311,7 @@ + (id)initWithLaunchOptions:(NSDictionary*)launchOptions
 
 #pragma mark - Define the rootview controller of the UINavigation-Stack
 + (UIViewController*)topViewController {
-  return [self topViewControllerWithRootViewController:[UIApplication sharedApplication].keyWindow.rootViewController];
+    return [self topViewControllerWithRootViewController:[UIApplication sharedApplication].keyWindow.rootViewController];
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:314]: Why did you add spaces here ?

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 6 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:95]: because to keep the reference of the tags inside the blocks, I've define global variable here.
and if we define variable inside the blocks we will have to define something like this
`        __block NSMutableArray* subscriptionTags = [NSMutableArray arrayWithArray:[userDefaults arrayForKey:@"CleverPush_SUBSCRIPTION_TAGS"]];
`
and that `__block NSMutableArray` and block not retain the values of the tags which we have added during the api calls. It is always allocates its memory reference and keep the only one object except the bunch of data.

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 8 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:314]: @gbhrdt I just reindent the code.

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 9]
```diff
@@ -1462,21 +1464,65 @@ + (void)handleJSONNSURLResponse:(NSURLResponse*) response data:(NSData*) data er
...
++ (void)addSubscriptionTags:(NSArray*)tagIds {
+    [self addSubscriptionTags:tagIds callback:nil];
+}
+
++ (void)removeSubscriptionTags:(NSArray*)tagIds {
+    [self removeSubscriptionTags:tagIds callback:nil];
+}
+
++ (void)addSubscriptionTags:(NSArray*)tagIds callback:(void(^)(NSArray *))callback {
+    dispatch_group_t group = dispatch_group_create();
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1476]: What happens if I call this method in the main thread?
Will it block?

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 10]
```diff
@@ -1218,7 +1220,7 @@ + (void)handleNotificationOpened:(NSDictionary*)payload isActive:(BOOL)isActive
     NSLog(@"CleverPush: handleNotificationOpened, %@, %@", action, payload);
     
     [CleverPush setNotificationClicked:notificationId withChannelId:[payload valueForKeyPath:@"channel._id"] withSubscriptionId:[payload valueForKeyPath:@"subscription._id"] withAction:action];
-
+
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1223]: remove this line

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 11]
```diff
@@ -1462,21 +1464,65 @@ + (void)handleJSONNSURLResponse:(NSURLResponse*) response data:(NSData*) data er
...
+    }
+    dispatch_group_notify(group, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
+        if (callback) {
+            callback([self getSubscriptionTags]);
+        }
+    });
+}
+
++ (void)removeSubscriptionTags:(NSArray*)tagIds callback:(void(^)(NSArray *))callback{
+    dispatch_group_t group = dispatch_group_create();
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1491]: What happens if I call this method in the main thread?
Will it block?

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 6 above)
Comment (gbhrdt) [reply CleverPush/CleverPush/Source/CleverPush.m:95]: I am not happy with this solution.

It still might happen that a user calls "addSubscriptionTag" many times at once. I guess a global variable will cause issues and race conditions here.

Comment (gbhrdt) [reply CleverPush/CleverPush/Source/CleverPush.m:95]: Please find a way to keep this in the method scope

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/CleverPush/Source/CleverPush.m [hunk 12]
```diff
@@ -1495,17 +1541,23 @@ + (void)addSubscriptionTag:(NSString*)tagId {
...
+                }
+                
             } onFailure:nil];
         });
     }];
 }
 
 #pragma mark - Remove subscription tag by calling api. subscription/untag
-+ (void)removeSubscriptionTag:(NSString*)tagId {
++ (void)removeSubscriptionTag:(NSString*)tagId callBack:(void (^)(NSString *))callback {
```
Comment (gbhrdt) [inline CleverPush/CleverPush/Source/CleverPush.m:1558]: callBack

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 9 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:1476]: Nope it will not block.

File: CleverPush/CleverPush/Source/CleverPush.m (hunk 11 above)
Comment (Azharhussain20) [reply CleverPush/CleverPush/Source/CleverPush.m:1490]: Nope it will not block.
I have checked this scenario by putting this method on main thread.
